tuf/client_rework/metadata_bundle.py

```diff
@@ -0,0 +1,505 @@
+    unique_keys = set()
+    for keyid in role["keyids"]:
+        key_metadata = keys[keyid]
+        key, dummy = sslib_keys.format_metadata_to_key(key_metadata)
+
+        try:
+            if unverified.verify(key):
+                unique_keys.add(key["keyval"]["public"])
+        except Exception as e:  # pylint: disable=broad-except
+            # TODO specify the Exceptions (see issue #1351)
+            logger.info("verify failed: %s", e)
+
+    return len(unique_keys) >= role["threshold"]
+
+
+class MetadataBundle(abc.Mapping):
+    """Internal class to keep track of valid metadata in Updater
+
+    MetadataBundle ensures that metadata is valid. It provides easy ways to
+    update the metadata with the caller making decisions on what is updated.
+    """
+
+    def __init__(self, repository_path: str):
+        """Initialize by loading root metadata from disk"""
+        self._path = repository_path
+        self._bundle = {}  # type: Dict[str: Metadata]
+        self.reference_time = None
+
+        # Load and validate the local root metadata
+        # Valid root metadata is required
+        logger.debug("Loading local root")
+        try:
+            with open(os.path.join(self._path, "root.json"), "rb") as f:
+                self._load_intermediate_root(f.read())
+        except (OSError, exceptions.RepositoryError) as e:
+            raise exceptions.RepositoryError(
+                "Failed to load local root metadata"
+            ) from e
+
+    # Implement Mapping
+    def __getitem__(self, key: str) -> Metadata:
+        return self._bundle[key]
+
+    def __len__(self) -> int:
+        return len(self._bundle)
+
+    def __iter__(self) -> Iterator[Metadata]:
+        return iter(self._bundle)
+
+    # Helper properties for top level metadata
+    @property
+    def root(self) -> Optional[Metadata]:
+        return self._bundle.get("root")
+
+    @property
+    def timestamp(self) -> Optional[Metadata]:
+        return self._bundle.get("timestamp")
+
+    @property
+    def snapshot(self) -> Optional[Metadata]:
+        return self._bundle.get("snapshot")
+
+    @property
+    def targets(self) -> Optional[Metadata]:
+        return self._bundle.get("targets")
+
+    # Public methods
+    def update_root(self, data: bytes):
+        """Update root metadata with data from remote repository."""
+        logger.debug("Updating root")
+
+        self._load_intermediate_root(data)
+        self.root.to_file(os.path.join(self._path, "root.json"))
+
+    def root_update_finished(self):
+        """Mark root metadata as final."""
+        if self.reference_time is not None:
+            raise ValueError("Root update is already finished")
+
+        # Store our reference "now", verify root expiry
+        self.reference_time = datetime.utcnow()
+        if self.root.signed.is_expired(self.reference_time):
+            raise exceptions.ExpiredMetadataError("New root.json is expired")
```

Comment:
Maybe yes. The spec does say that...

This choice made sense to me as it is (very close to) the time that client last checks for root.json -- so theoretically  there's actual value in comparing it to e.g. server publish times if issues are debugged...

But this could be a situation where following the spec to letter is a better choice. That just means I need an extra variable for `root_is_final` :)

Reply:
I wanted to suggest adding an additional flag marking `root_update_finished`. I see that reference time serves exactly like that but something like `if root_update_finished:` is instantly comprehensible.